official/coin_build/man/man3/SbString.3
.TH "SbString" 3 "Sun May 28 2017" "Version 4.0.0a" "Coin" \" -*- nroff -*-
.ad l
.nh
.SH NAME
SbString \- The \fBSbString\fP class is a string class with convenience functions for string operations\&.
.PP
This is the class used for storing and working with character strings\&. It automatically takes care of supporting all the 'bookkeeping' tasks usually associated with working with character strings, like memory allocation and deallocation etc\&.  

.SH SYNOPSIS
.br
.PP
.PP
\fC#include <Inventor/SbString\&.h>\fP
.SS "Public Member Functions"

.in +1c
.ti -1c
.RI "\fBSbString\fP (void)"
.br
.ti -1c
.RI "\fBSbString\fP (const char *s)"
.br
.ti -1c
.RI "\fBSbString\fP (const wchar_t *s)"
.br
.ti -1c
.RI "\fBSbString\fP (const char *s, \fBint\fP start, \fBint\fP end)"
.br
.ti -1c
.RI "\fBSbString\fP (const \fBSbString\fP &s)"
.br
.ti -1c
.RI "\fBSbString\fP (const \fBint\fP digits)"
.br
.ti -1c
.RI "\fB~SbString\fP ()"
.br
.ti -1c
.RI "uint32_t \fBhash\fP (void) const"
.br
.ti -1c
.RI "\fBint\fP \fBgetLength\fP (void) const"
.br
.ti -1c
.RI "void \fBmakeEmpty\fP (SbBool freeold=TRUE)"
.br
.ti -1c
.RI "const char * \fBgetString\fP (void) const"
.br
.ti -1c
.RI "\fBSbString\fP \fBgetSubString\fP (\fBint\fP startidx, \fBint\fP endidx=\-1) const"
.br
.ti -1c
.RI "void \fBdeleteSubString\fP (\fBint\fP startidx, \fBint\fP endidx=\-1)"
.br
.ti -1c
.RI "void \fBaddIntString\fP (const \fBint\fP value)"
.br
.ti -1c
.RI "char \fBoperator[]\fP (\fBint\fP index) const"
.br
.ti -1c
.RI "\fBSbString\fP & \fBoperator=\fP (const char *s)"
.br
.ti -1c
.RI "\fBSbString\fP & \fBoperator=\fP (const \fBSbString\fP &s)"
.br
.ti -1c
.RI "\fBSbString\fP & \fBoperator+=\fP (const char *s)"
.br
.ti -1c
.RI "\fBSbString\fP & \fBoperator+=\fP (const \fBSbString\fP &s)"
.br
.ti -1c
.RI "\fBSbString\fP & \fBoperator+=\fP (const char c)"
.br
.ti -1c
.RI "\fBint\fP \fBoperator!\fP (void) const"
.br
.ti -1c
.RI "\fBint\fP \fBcompareSubString\fP (const char *text, \fBint\fP offset=0) const"
.br
.ti -1c
.RI "\fBSbString\fP & \fBsprintf\fP (const char *formatstr,\&.\&.\&.)"
.br
.ti -1c
.RI "\fBSbString\fP & \fBvsprintf\fP (const char *formatstr, va_list args)"
.br
.ti -1c
.RI "void \fBapply\fP (char(*func)(char input))"
.br
.ti -1c
.RI "\fBint\fP \fBfind\fP (const \fBSbString\fP &s) const"
.br
.ti -1c
.RI "SbBool \fBfindAll\fP (const \fBSbString\fP &s, \fBSbIntList\fP &found) const"
.br
.ti -1c
.RI "\fBSbString\fP \fBlower\fP () const"
.br
.ti -1c
.RI "\fBSbString\fP \fBupper\fP () const"
.br
.ti -1c
.RI "void \fBprint\fP (std::FILE *fp) const"
.br
.in -1c
.SS "Static Public Member Functions"

.in +1c
.ti -1c
.RI "static uint32_t \fBhash\fP (const char *s)"
.br
.in -1c
.SS "Friends"

.in +1c
.ti -1c
.RI "\fBint\fP \fBoperator==\fP (const \fBSbString\fP &sbstr, const char *s)"
.br
.ti -1c
.RI "\fBint\fP \fBoperator==\fP (const char *s, const \fBSbString\fP &sbstr)"
.br
.ti -1c
.RI "\fBint\fP \fBoperator==\fP (const \fBSbString\fP &str1, const \fBSbString\fP &str2)"
.br
.ti -1c
.RI "\fBint\fP \fBoperator!=\fP (const \fBSbString\fP &sbstr, const char *s)"
.br
.ti -1c
.RI "\fBint\fP \fBoperator!=\fP (const char *s, const \fBSbString\fP &sbstr)"
.br
.ti -1c
.RI "\fBint\fP \fBoperator!=\fP (const \fBSbString\fP &str1, const \fBSbString\fP &str2)"
.br
.ti -1c
.RI "\fBint\fP \fBoperator<\fP (const \fBSbString\fP &sbstr, const char *s)"
.br
.ti -1c
.RI "\fBint\fP \fBoperator<\fP (const char *s, const \fBSbString\fP &sbstr)"
.br
.ti -1c
.RI "\fBint\fP \fBoperator<\fP (const \fBSbString\fP &str1, const \fBSbString\fP &str2)"
.br
.ti -1c
.RI "\fBint\fP \fBoperator>\fP (const \fBSbString\fP &sbstr, const char *s)"
.br
.ti -1c
.RI "\fBint\fP \fBoperator>\fP (const char *s, const \fBSbString\fP &sbstr)"
.br
.ti -1c
.RI "\fBint\fP \fBoperator>\fP (const \fBSbString\fP &str1, const \fBSbString\fP &str2)"
.br
.ti -1c
.RI "const \fBSbString\fP \fBoperator+\fP (const \fBSbString\fP &str1, const \fBSbString\fP &str2)"
.br
.ti -1c
.RI "const \fBSbString\fP \fBoperator+\fP (const \fBSbString\fP &sbstr, const char *s)"
.br
.ti -1c
.RI "const \fBSbString\fP \fBoperator+\fP (const char *s, const \fBSbString\fP &sbstr)"
.br
.in -1c
.SH "Detailed Description"
.PP 
The \fBSbString\fP class is a string class with convenience functions for string operations\&.
.PP
This is the class used for storing and working with character strings\&. It automatically takes care of supporting all the 'bookkeeping' tasks usually associated with working with character strings, like memory allocation and deallocation etc\&. 

This class should also be well suited for use by the application programmer throughout the application using the Coin library\&.
.PP
\fBSee also:\fP
.RS 4
\fBSbName\fP 
.RE
.PP

.SH "Constructor & Destructor Documentation"
.PP 
.SS "SbString::SbString (void)\fC [inline]\fP"
This is the default constructor\&. It initializes the string to be empty\&. 
.SS "SbString::SbString (const char * str)\fC [inline]\fP"
This constructor assigns from the given string\&. 
.SS "SbString::SbString (const wchar_t * str)\fC [inline]\fP"
This constructor converts the given string to UTF-8 and assigns it\&. 
.SS "SbString::SbString (const char * str, \fBint\fP start, \fBint\fP end)\fC [inline]\fP"
This constructor constructs a string from the given substring from \fIstart\fP to \fIend\fP indices inclusive\&. If \fIend\fP is -1, the substring from \fIstart\fP until the end of the string is used\&. 
.SS "SbString::SbString (const \fBSbString\fP & str)\fC [inline]\fP"
This is the copy constructor\&. 
.SS "SbString::SbString (const \fBint\fP digits)\fC [inline]\fP"
Construct an \fBSbString\fP instance containing the \fIdigits\fP of the integer argument\&. 
.SS "SbString::~SbString ()\fC [inline]\fP"
The destructor\&. Deallocates any internal resources used during the lifetime of the \fBSbString\fP instance\&. 
.SH "Member Function Documentation"
.PP 
.SS "uint32_t SbString::hash (void) const\fC [inline]\fP"
This method returns a reasonable hash value for the current string\&.
.PP
\fBSee also:\fP
.RS 4
uint32_t \fBSbString::hash(const char * s)\fP 
.RE
.PP

.SS "uint32_t SbString::hash (const char * s)\fC [inline]\fP, \fC [static]\fP"
This static method returns a hash value for the given string\&. 
.SS "\fBint\fP SbString::getLength (void) const\fC [inline]\fP"
This method returns the length of the string\&. 
.SS "void SbString::makeEmpty (SbBool freeold = \fCTRUE\fP)\fC [inline]\fP"
This method clears the string, making it an empty string ('')\&. If \fIfreeold\fP is \fCTRUE\fP (which is the default), the memory used by the old string is freed\&. Otherwise, memory will be kept and reused when the string is manipulated later\&. 
.SS "const char * SbString::getString (void) const\fC [inline]\fP"
This method returns the pointer to the string (character array)\&. 
.SS "\fBSbString\fP SbString::getSubString (\fBint\fP startidx, \fBint\fP endidx = \fC\-1\fP) const\fC [inline]\fP"
This method returns a new string which contains a substring defined by the given indices \fIstartidx\fP and \fIendidx\fP (inclusive)\&.
.PP
If \fIendidx\fP is -1, the substring from \fIstartidx\fP to the end of the string is used\&.
.PP
This will return a string which is (\fIendidx\fP - \fIstartidx\fP + 1) characters long, i\&.e\&. if this string is 'foo/bar' and we call by SbString::getSubString(0, 3), the returned string will be 'foo/'\&. 
.SS "void SbString::deleteSubString (\fBint\fP startidx, \fBint\fP endidx = \fC\-1\fP)\fC [inline]\fP"
This method deletes the substring defined by \fIstartidx\fP and \fIendidx\fP (inclusive)\&. If \fIendidx\fP is -1, the substring from \fIstartidx\fP to the end of the string is deleted\&. 
.SS "void SbString::addIntString (const \fBint\fP value)\fC [inline]\fP"
Constructs a string from the given integer (e\&.g\&. intToString(42) creates the string '42'), and adds this to the contents of the string\&. 
.SS "char SbString::operator[] (\fBint\fP index) const\fC [inline]\fP"
Returns character at position \fIindex\fP in the string\&.
.PP
\fBSee also:\fP
.RS 4
\fBgetSubString()\fP 
.RE
.PP

.SS "\fBSbString\fP & SbString::operator= (const char * str)\fC [inline]\fP"
This is the assignment operator\&. 
.SS "\fBSbString\fP & SbString::operator= (const \fBSbString\fP & str)\fC [inline]\fP"
Assign from the given string\&. 
.SS "\fBSbString\fP & SbString::operator+= (const char * str)\fC [inline]\fP"
Concatenate the given string to the end of the current one\&. 
.SS "\fBSbString\fP & SbString::operator+= (const \fBSbString\fP & str)\fC [inline]\fP"
Concatenate the given string to the end of the current one\&. 
.SS "\fBSbString\fP & SbString::operator+= (const char c)\fC [inline]\fP"
Concatenate the given character to the end of the current string\&.
.PP
\fBNote:\fP
.RS 4
This member function is not compatible with OpenInventor\&. 
.RE
.PP

.SS "\fBint\fP SbString::operator! (void) const\fC [inline]\fP"
This unary operator results in \fCTRUE\fP if the current string is empty ('') or \fCFALSE\fP otherwise\&. 
.SS "\fBint\fP SbString::compareSubString (const char * text, \fBint\fP offset = \fC0\fP) const\fC [inline]\fP"
Uses \fItext\fP as a substring to be compared to the string, starting at offset \fIoffset\fP\&. Return value is 0 upon success, and the failing character comparison difference upon failure\&. 
.SS "\fBSbString\fP & SbString::sprintf (const char * formatstr,  \&.\&.\&.)\fC [inline]\fP"
Set \fBSbString\fP instance to the formatted string \fIformatstr\fP, replacing the current contents\&. The control characters within \fIformatstr\fP and the remaining arguments should follow the conventions of the printf() call\&.
.PP
Note that this function is not part of the original Open Inventor API\&. 
.SS "\fBSbString\fP & SbString::vsprintf (const char * formatstr, va_list args)\fC [inline]\fP"
Set \fBSbString\fP instance to the formatted string \fIformatstr\fP, replacing the current contents\&. The control characters within \fIformatstr\fP and the arguments of the \fIargs\fP argument list should follow the conventions of the printf() call\&.
.PP
Note that this function is not part of the original Open Inventor API\&. 
.SS "\fBint\fP SbString::find (const \fBSbString\fP & strarg) const"
If \fIs\fP is found, the method returns the first index where \fIs\fP starts\&. Otherwise it returns -1\&.
.PP
Note: \fBSbString::find()\fP is a Coin specific extension to the original Open Inventor API\&.
.PP
\fBSee also:\fP
.RS 4
\fBSbString::findAll()\fP 
.RE
.PP
\fBSince:\fP
.RS 4
Coin 2\&.0 
.RE
.PP

.SS "SbBool SbString::findAll (const \fBSbString\fP & strarg, \fBSbIntList\fP & found) const"
All occurences of \fIstr\fP is represented in \fIfound\fP as indices to the characters where \fIstr\fP starts\&. If 1 or more is found, \fCTRUE\fP is returned, else \fCFALSE\fP is returned\&.
.PP
Note: \fBSbString::findAll()\fP is an extension to the original Open Inventor API\&.
.PP
\fBSee also:\fP
.RS 4
\fBSbString::find()\fP 
.RE
.PP
\fBSince:\fP
.RS 4
Coin 2\&.0 
.RE
.PP

.SS "\fBSbString\fP SbString::lower () const"
Converts all of the characters to lowercase using tolower()\&.
.PP
\fBSince:\fP
.RS 4
Coin 3\&.1 
.RE
.PP

.SS "\fBSbString\fP SbString::upper () const"
Converts all of the characters to uppercase using toupper()\&.
.PP
\fBSince:\fP
.RS 4
Coin 3\&.1 
.RE
.PP

.SS "void SbString::print (std::FILE * fp) const"
Dump the state of this object to the \fIfile\fP stream\&. Only works in debug version of library, method does nothing in an optimized compile\&. 
.SH "Friends And Related Function Documentation"
.PP 
.SS "\fBint\fP operator== (const \fBSbString\fP & str, const char * s)\fC [friend]\fP"
Equality operator\&. Check if the strings have the same contents\&. 
.SS "\fBint\fP operator== (const char * s, const \fBSbString\fP & str)\fC [friend]\fP"
Equality operator\&. Check if the strings have the same contents\&. 
.SS "\fBint\fP operator== (const \fBSbString\fP & str1, const \fBSbString\fP & str2)\fC [friend]\fP"
Equality operator\&. Check if the strings have the same contents\&. 
.SS "\fBint\fP \fBoperator!\fP= (const \fBSbString\fP & str, const char * s)\fC [friend]\fP"
Inequality operator\&. 
.SS "\fBint\fP \fBoperator!\fP= (const char * s, const \fBSbString\fP & str)\fC [friend]\fP"
Inequality operator\&. 
.SS "\fBint\fP \fBoperator!\fP= (const \fBSbString\fP & str1, const \fBSbString\fP & str2)\fC [friend]\fP"
Inequality operator\&. 
.SS "\fBint\fP operator< (const \fBSbString\fP & str, const char * s)\fC [friend]\fP"
Less than operator\&. Results in \fCTRUE\fP if the first character that does not match has a lower value in \fIstr\fP than in \fIs\fP\&. 
.SS "\fBint\fP operator< (const char * s, const \fBSbString\fP & str)\fC [friend]\fP"
Less than operator\&. Results in \fCTRUE\fP if the first character that does not match has a lower value in \fIs\fP than in \fIstr\fP\&. 
.SS "\fBint\fP operator< (const \fBSbString\fP & str1, const \fBSbString\fP & str2)\fC [friend]\fP"
Less than operator\&. Results in \fCTRUE\fP if the first character that does not match has a lower value in \fIstr1\fP than in \fIstr2\fP\&. 
.SS "\fBint\fP operator> (const \fBSbString\fP & str, const char * s)\fC [friend]\fP"
Greater than operator\&. Results in \fCTRUE\fP if the first character that does not match has a higher value in \fIstr\fP than in \fIs\fP\&. 
.SS "\fBint\fP operator> (const char * s, const \fBSbString\fP & str)\fC [friend]\fP"
Greater than operator\&. Results in \fCTRUE\fP if the first character that does not match has a higher value in \fIs\fP than in \fIstr\fP\&. 
.SS "\fBint\fP operator> (const \fBSbString\fP & str1, const \fBSbString\fP & str2)\fC [friend]\fP"
Greater than operator\&. Results in \fCTRUE\fP if the first character that does not match has a higher value in \fIstr1\fP than in \fIstr2\fP\&. 
.SS "const \fBSbString\fP operator+ (const \fBSbString\fP & str1, const \fBSbString\fP & str2)\fC [friend]\fP"
Addition operator\&. 
.SS "const \fBSbString\fP operator+ (const \fBSbString\fP & sbstr, const char * s)\fC [friend]\fP"
Addition operator\&. 
.SS "const \fBSbString\fP operator+ (const char * s, const \fBSbString\fP & sbstr)\fC [friend]\fP"
Addition operator\&. 

.SH "Author"
.PP 
Generated automatically by Doxygen for Coin from the source code\&.


insanity/coin/src/base/SbString.cpp
/**************************************************************************\
 * Copyright (c) Kongsberg Oil & Gas Technologies AS
 * All rights reserved.
 * 
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are
 * met:
 * 
 * Redistributions of source code must retain the above copyright notice,
 * this list of conditions and the following disclaimer.
 * 
 * Redistributions in binary form must reproduce the above copyright
 * notice, this list of conditions and the following disclaimer in the
 * documentation and/or other materials provided with the distribution.
 * 
 * Neither the name of the copyright holder nor the names of its
 * contributors may be used to endorse or promote products derived from
 * this software without specific prior written permission.
 * 
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
 * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
 * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
 * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
 * HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
 * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
 * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
 * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
 * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
 * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
\**************************************************************************/

/*!
  \class SbString SbString.h Inventor/SbString.h
  \brief The SbString class is a string class with convenience functions for string operations.
  \ingroup base

  This is the class used for storing and working with character
  strings. It automatically takes care of supporting all the
  "bookkeeping" tasks usually associated with working with character
  strings, like memory allocation and deallocation etc.

  This class should also be well suited for use by the application
  programmer throughout the application using the Coin library.

  \sa SbName
*/

#include <Inventor/SbString.h>
#include <cctype>

#include <Inventor/lists/SbList.h>
#include "coindefs.h"

// FIXME: translate the find() and findAll() methods to C (will first
// need a growable array in C to handle the second argument to
// findAll()). 20020522 mortene.

// Helper function for find() and findAll().
static void
compute_prefix_function(SbList <int> & pi, const SbString & str)
{
  int len = str.getLength();
  pi.append(0);
  int k = 0;

  for (int q = 1; q < len; q++){
    while(k > 0 && (str.operator[](k) != str.operator[](q)))
      k = pi[k];
    if (str.operator[](k) == str.operator[](q))
      k++;
    pi.append(k);
  }
}

/*
   This is a reimplementation of sdbm hash algorithm from the sdbm project
*/
static unsigned int hash(const unsigned char * str, int n)
{
  unsigned long hash = 0;

  while (n--) {
    hash = (*str++) + (hash << 6) + (hash << 16) - hash;
  }

  return hash;
}

unsigned int SbHashFunc(const SbString & key) {
  const unsigned char * cKey = reinterpret_cast<const unsigned char *>(key.getString());
  return hash(cKey,key.getLength());
}


/*!
  If \a s is found, the method returns the first index where \a s
  starts. Otherwise it returns -1.

  Note: SbString::find() is a Coin specific extension to the original
  Open Inventor API.

  \sa SbString::findAll()
  \since Coin 2.0
*/
int
SbString::find(const SbString & strarg) const
{
  int lenthis = this->getLength();
  int lenstr = strarg.getLength();

  if (!lenthis) return -1;
  if (lenstr > lenthis) return -1;

  SbList<int> pi;
  compute_prefix_function(pi, strarg);
  int q = 0;

  for (int i = 0; i < lenthis; i ++){
    while (q > 0 && (strarg.operator[](q) != this->operator[](i)))
      q = pi.operator[](q - 1);
    if (strarg.operator[](q) == this->operator[](i))
      q++;
    if (q == lenstr){
      return (i - (lenstr - 1));
    }
  }
  return -1;
}

/*!
  All occurences of \a str is represented in \a found as indices to
  the characters where \a str starts. If 1 or more is found, \c true
  is returned, else \c false is returned.

  Note: SbString::findAll() is an extension to the original Open
  Inventor API.

  \sa SbString::find()
  \since Coin 2.0
*/

bool
SbString::findAll(const SbString & strarg, SbIntList & found) const
{
  // The KMP string matching algorithm is used for this method
  int lenthis = this->getLength();
  int lenstr = strarg.getLength();
  found.truncate(0);

  if (!lenthis) return false;
  if (lenstr > lenthis) return false;

  SbList<int> pi;
  compute_prefix_function(pi, strarg);
  int q = 0;

  for (int i = 0; i < lenthis; i ++){
    while (q > 0 && (strarg.operator[](q) != this->operator[](i)))
      q = pi.operator[](q - 1);
    if (strarg.operator[](q) == this->operator[](i))
      q++;
    if (q == lenstr){
      found.append(i - (lenstr - 1));
      q = pi.operator[](q - 1);
    }
  }
  if (!found.getLength()) return false;
  return true;
}

/*!
  Converts all of the characters to lowercase using tolower().

  \since Coin 3.1
*/
SbString
SbString::lower() const
{
  int lenthis = this->getLength();
  SbString newstr;

  for (int i = 0; i < lenthis; i++) {
    newstr += std::tolower((*this)[i]);
  }
  return newstr;
}

/*!
  Converts all of the characters to uppercase using toupper().

  \since Coin 3.1
*/
SbString
SbString::upper() const
{
  int lenthis = this->getLength();
  SbString newstr;

  for (int i = 0; i < lenthis; i++) {
    newstr += std::toupper((*this)[i]);
  }
  return newstr;
}

/*!
  Dump the state of this object to the \a file stream. Only works in
  debug version of library, method does nothing in an optimized
  compile.
*/
void
SbString::print(std::FILE * fp) const
{
#if COIN_DEBUG
  std::fputs(this->getString(),fp);
#endif // COIN_DEBUG
}


/*!
  \fn SbString::SbString(void)
  This is the default constructor.  It initializes the string to be empty.
*/

/*!
  \fn SbString::SbString(const SbString & str)
  This is the copy constructor.
*/

/*!
  \fn SbString::SbString(const char * str)
  This constructor assigns from the given string.
*/

/*!
  \fn SbString::SbString(const char * str, int start, int end)

  This constructor constructs a string from the given substring from
  \a start to \a end indices inclusive. If \a end is -1, the substring
  from \a start until the end of the string is used.
*/


/*!
  \fn SbString::SbString(const int digits)

  Construct an SbString instance containing the \a digits of the
  integer argument.
 */

/*!
  \fn void SbString::addIntString(const int value)

  Constructs a string from the given integer (e.g. intToString(42)
  creates the string "42"), and adds this to the contents of the
  string.
*/

/*!
  \fn int SbString::compareSubString(const char * text, int offset = 0) const

  Uses \a text as a substring to be compared to the string, starting at
  offset \a offset.  Return value is 0 upon success, and the failing
  character comparison difference upon failure.
*/

/*!
  \fn SbString SbString::getSubString(int startidx, int endidx) const

  This method returns a new string which contains a substring defined
  by the given indices \a startidx and \a endidx (inclusive).

  If \a endidx is -1, the substring from \a startidx to the end of the
  string is used.

  This will return a string which is (\a endidx - \a startidx + 1)
  characters long, i.e. if this string is "foo/bar" and we call by
  SbString::getSubString(0, 3), the returned string will be "foo/".
*/

/*!
  \fn void SbString::deleteSubString(int startidx, int endidx)

  This method deletes the substring defined by \a startidx and \a endidx
  (inclusive).  If \a endidx is -1, the substring from \a startidx to the
  end of the string is deleted.
*/

/*!
  \fn SbString::~SbString()

  The destructor. Deallocates any internal resources used during the
  lifetime of the SbString instance.
*/

/*!
  \fn uint32_t SbString::hash(void) const

  This method returns a reasonable hash value for the current string.

  \sa uint32_t SbString::hash(const char * s)
*/

/*!
  \fn uint32_t SbString::hash(const char * s)

  This static method returns a hash value for the given string.
*/

/*!
  \fn int SbString::getLength(void) const

  This method returns the length of the string.
*/

/*!
  \fn const char * SbString::getString(void) const

  This method returns the pointer to the string (character array).
*/

/*!
  \fn void SbString::makeEmpty(bool freeold)

  This method clears the string, making it an empty string ("").
  If \a freeold is \c true (which is the default), the memory used by
  the old string is freed.  Otherwise, memory will be kept and reused
  when the string is manipulated later.
*/

/*!
  \fn char SbString::operator[](int index) const

  Returns character at position \a index in the string.

  \sa getSubString()
*/

/*!
  \fn SbString & SbString::operator=(const char * str)

  This is the assignment operator.
*/

/*!
  \fn SbString & SbString::operator=(const SbString & str)

  Assign from the given string.
*/

/*!
  \fn SbString & SbString::operator+=(const char * str)

  Concatenate the given string to the end of the current one.
*/

/*!
  \fn SbString & SbString::operator+=(const SbString & str)

  Concatenate the given string to the end of the current one.
*/

/*!
  \fn SbString & SbString::operator+=(const char c)

  Concatenate the given character to the end of the current string.

  \note This member function is not compatible with OpenInventor.
*/

/*!
  \fn int SbString::operator!(void) const

  This unary operator results in \c true if the current string is empty ("")
  or \c false otherwise.
*/

/*!
  \fn bool operator==(const SbString & str, const char * s)

  Equality operator. Check if the strings have the same contents.

  \relates SbString
*/

/*!
  \fn bool operator==(const char * s, const SbString & str)

  Equality operator. Check if the strings have the same contents.

  \relates SbString
*/

/*!
  \fn bool operator==(const SbString & str1, const SbString & str2)

  Equality operator. Check if the strings have the same contents.

  \relates SbString
*/

/*!
  \fn bool operator!=(const SbString & str, const char * s)

  Inequality operator.

  \relates SbString
*/

/*!
  \fn bool operator!=(const char * s, const SbString & str)

  Inequality operator.

  \relates SbString
*/

/*!
  \fn bool operator!=(const SbString & str1, const SbString & str2)

  Inequality operator.

  \relates SbString
*/

/*!
  \fn const SbString operator+(const SbString & str1, const SbString & str2)

  Addition operator.

  \relates SbString
*/

/*!
  \fn const SbString operator+(const SbString & sbstr, const char * s)

  Addition operator.

  \relates SbString
*/

/*!
  \fn const SbString operator+(const char * s, const SbString & sbstr)

  Addition operator.

  \relates SbString
*/

/*!
  \fn SbString & SbString::sprintf(const char * formatstr, ...)

  Set SbString instance to the formatted string \a formatstr,
  replacing the current contents.  The control characters within \a
  formatstr and the remaining arguments should follow the conventions
  of the printf() call.

  Note that this function is not part of the original Open Inventor
  API.
*/

/*!
  \fn SbString & SbString::vsprintf(const char * formatstr, va_list args)

  Set SbString instance to the formatted string \a formatstr,
  replacing the current contents.  The control characters within \a
  formatstr and the arguments of the \a args argument list should
  follow the conventions of the printf() call.

  Note that this function is not part of the original Open Inventor
  API.
*/

#ifdef COIN_TEST_SUITE
#include <Inventor/SbString.h>

static void * createInstance(void)
{
  return (void *)0x1234;
}

BOOST_AUTO_TEST_CASE(testAddition)
{
  SbString str1("First");
  SbString str2("Second");
  const char *cstr1 = "Erste";
  const char *cstr2 = "Zweite";

  SbString a = str1 + str2;
  SbString b = cstr1 + str2;
  SbString c = str1 + cstr2;

  BOOST_CHECK_MESSAGE(a == SbString("FirstSecond"),
                      "operator+ error");
  BOOST_CHECK_MESSAGE(b == SbString("ErsteSecond"),
                      "operator+ error");
  BOOST_CHECK_MESSAGE(c == SbString("FirstZweite"),
                      "operator+ error");
}

#endif // COIN_TEST_SUITE
